File: C++/Problems/Problems/problem_45.cpp
#include "problem_45.h"

bool problem_45::test() {
    std::vector nums { 2, 3, 1, 1, 4 };

    int expected = 2;

    auto result = jump(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_45::jump(std::vector<int> &nums) {
    int n = nums.size();
    int maxReachable = 0;
    int curPos = 0;
    int jumps = 0;
    for (int i = 0; i < n; ++i)
    {
        if (i > curPos)
        {
            jumps++;
            curPos = maxReachable;
        }
        maxReachable = std::max(maxReachable, i + nums[i]);
    }

    return jumps;
}

File: Python/Problems/Problems/problem45.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem45(Problem):
    def test(self) -> bool:
        nums = [2, 3, 1, 1, 4]

        expected = 2

        result = self.jump(nums)

        print(result)

        return result == expected

    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        maxReachable = 0
        curPos = 0
        jumps = 0
        for i in range(n):
            if i > curPos:
                jumps += 1
                curPos = maxReachable
            maxReachable = max(maxReachable, i + nums[i])

        return jumps